src/simulations/local_elec.py

```diff
@@ -546,19 +548,53 @@ def policy_to_rewardvar(
         self.heat.next_T(update=True)
         self._check_constraints(
             bool_penalty, date, loads, E_req_only, h, last_step, home_vars)
-
+        
+        if self.prm['syst']['n_homesP'] > 0:
+            netp0 = self.prm['loads']['netp0'][:, h]
+        else:
+            netp0 = []
         if self.prm['grd']['manage_voltage']:
-            hourly_line_losses, voltage = self.network.pf_simulation(home_vars['netp'])
+            if self.prm['syst']['n_homesP'] > 0:
+                q_heat_home_car_non_flex = self._calculate_reactive_power(
+                    netp0,
+                    self.prm['grd']['pf_non_flex_heat_home_car']
+                    )
+            else:
+                q_heat_home_car_non_flex = []
+                q_heat_home_flex = self._calculate_reactive_power(
+                    home_vars['tot_cons'], self.prm['grd']['pf_flexible_heat_home'])
+                # q_car_flex will be a decision variable
+                p_car_flex = - (np.array(self.car.loss_ch) + np.array(self.car.charge)) \
+                    + np.array(self.car.discharge)
+                q_car_flex = self._calculate_reactive_power(
+                    p_car_flex, self.prm['grd']['pf_flexible_heat_home'])
+                # p_car_flex is needed to set apparent power limits
+                netq_flex = q_car_flex + q_heat_home_flex
+                netq_non_flex = q_heat_home_car_non_flex
+                # import/export external grid
+                q_ext_grid = sum(q_heat_home_car_non_flex) + sum(q_car_flex) \
+                    + sum(q_heat_home_flex)
+
+            hourly_line_losses, voltage = self.network.pf_simulation(
+                home_vars['netp'], netp0,
+                netq_flex, netq_non_flex)
             voltage_squared = np.square(voltage)
 
         else:
             voltage_squared = None
             hourly_line_losses = 0
+            q_ext_grid = 0
+
 
         if sum(bool_penalty) > 0:
             constraint_ok = False
 
-        return home_vars, loads, hourly_line_losses, voltage_squared, constraint_ok
+        return (home_vars, loads, hourly_line_losses, voltage_squared,
+                q_ext_grid, netp0, constraint_ok)
+
+    def _calculate_reactive_power(self, active_power, power_factor):
+        reactive_power = active_power * math.tan(math.acos(power_factor))
+        return reactive_power
```

Comment:
this could be in utils.py; it does not require any attribute or method from `env`. Then it is just a function that can be called from anywhere, this has no impact/interdependence with the `env` object.

Reply:
Yes, that makes sense. I've moved this function to userdeftools and now import and use it in the optimizer, explorer and local_elec.